src/TokenBuyer.sol

```diff
@@ -0,0 +1,308 @@
+    IOUToken public immutable iouToken;
+
+    /// @notice the contract used to fetch the price of `paymentToken` in ETH.
+    IPriceFeed public priceFeed;
+
+    /// @notice the minimum `paymentToken` balance this contract should aim to hold, in WAD format.
+    uint256 public baselinePaymentTokenAmount;
+
+    /// @notice the amount of basis points to increase `paymentToken` price by, to increase the incentive to transact with this contract.
+    uint16 public botIncentiveBPs;
+
+    constructor(
+        IERC20 _paymentToken,
+        uint8 _paymentTokenDecimals,
+        IOUToken _iouToken,
+        IPriceFeed _priceFeed,
+        uint256 _baselinePaymentTokenAmount,
+        uint16 _botIncentiveBPs,
+        address _owner
+    ) {
+        paymentToken = _paymentToken;
+        paymentTokenDecimals = _paymentTokenDecimals;
+        iouToken = _iouToken;
+
+        priceFeed = _priceFeed;
+        baselinePaymentTokenAmount = _baselinePaymentTokenAmount;
+        botIncentiveBPs = _botIncentiveBPs;
+        _transferOwnership(_owner);
+
+        // Using a temp variable because you can't set an immutable variable inside an if statement.
+        uint256 paymentTokenToWADFactorTemp = 1;
+        if (paymentTokenDecimals < WAD_DECIMALS) {
+            paymentTokenToWADFactorTemp = 10**(WAD_DECIMALS - paymentTokenDecimals);
+        } else if (paymentTokenDecimals > WAD_DECIMALS) {
+            paymentTokenToWADFactorTemp = 10**(paymentTokenDecimals - WAD_DECIMALS);
+        }
+
+        paymentTokenToWADFactor = paymentTokenToWADFactorTemp;
+    }
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      EXTERNAL TRANSACTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /**
+     * @notice Buy ETH from this contract in exchange for the ERC20 this token wants to acquire. The price
+     * is determined using `priceFeed` plus `botIncentiveBPs` basis points.
+     * @dev if `tokenAmount > tokenAmountNeeded()` uses the maximum amount possible. This function allows reentry because it does
+     * not allow double spending or exceeding the contract's {tokenAmountNeeded()}.
+     * @param tokenAmountWAD the amount of ERC20 tokens msg.sender wishes to sell to this contract in exchange for ETH, in WAD format.
+     */
+    function buyETH(uint256 tokenAmountWAD) external nonReentrant {
+        uint256 amount = min(tokenAmountWAD, tokenAmountNeeded());
+
+        paymentToken.safeTransferFrom(msg.sender, address(this), wadToTokenDecimals(amount));
+
+        safeSendETH(msg.sender, ethAmountPerTokenAmount(amount), '');
+    }
+
+    function buyETH(
+        uint256 tokenAmountWAD,
+        address to,
+        bytes calldata data
+    ) external nonReentrant {
+        uint256 amount = min(tokenAmountWAD, tokenAmountNeeded());
+        uint256 balanceBefore = paymentTokenBalance();
+
+        safeSendETH(to, ethAmountPerTokenAmount(amount), abi.encode(msg.sender, amount, data));
```

Comment:
maybe better to call a named function rather than encoding parameters manually

Reply:
let's discuss!

Reply:
we're now moving in the direction of an explicit function, so resolving this thread